Grand circuit pathplanning test

GearBoxFox opened the pull request.

GearBoxFox committed:
Fix: better middle score auto

GearBoxFox committed:
Hey more things

Grandcircuit424 committed:
Feat: Switched DutyCycle encoders to CanCoders.
Made 6 paths

GearBoxFox committed:
Fix: removed normalization on CANCoder

Grandcircuit424 committed:
Fix: Path files

Grandcircuit424 committed:
Merge branch 'GrandCircuit-PathplanningTest' of https://github.com/TitaniumTitans/2023ChargedUp into GrandCircuit-PathplanningTest

GearBoxFox committed:
Merge branch 'main' into Gearfox-DriveBindings

GearBoxFox committed:
Merge branch 'main' into Gearfox-DriveBindings

GearBoxFox committed:
Feat: switch origin for auto

GearBoxFox committed:
Feat: More tag origin setting

Zoarial94 commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -146,7 +146,7 @@ public static class ArmConstants {
         public static final double ARM_OFFSET = 84 + 120 + 243;
 
-        public static final int ENCODER_PORT = 1;
+        public static final int ENCODER_PORT = 24;
```
What encoder more specifically?

GearBoxFox committed:
Fix: adjusted camera angle

GearBoxFox committed:
Revert commit "Feat: Through bore encoder to CANCoder"

GearBoxFox committed:
Feat: Fix for origin flipping

GearBoxFox committed:
Merge branch 'main' into GrandCircuit-PathplanningTest

# Conflicts:
#	src/main/java/frc/robot/Constants.java
#	src/main/java/frc/robot/subsystems/vision/CameraSubsystem.java

GearBoxFox committed:
Feat: removed unused path files and renamed them

GearBoxFox committed:
Merge branch 'main' into GrandCircuit-PathplanningTest

GearBoxFox committed:
Feat: Auto selector added mobility option

Zoarial94 committed:
ArmAngleSubsystem.java cleanup. Added explicit brake/coast functions

Zoarial94 committed:
ArmSupersystem.java cleanup

Zoarial94 committed:
Robot.java cleanup

Zoarial94 committed:
WristSubsystem.java cleanup

Grandcircuit424 committed:
Feat: Added auto for Balance(Left, Middle, Right)

Grandcircuit424 committed:
Feat: Added auto for PickUp(Right, Middle, Left)

GearBoxFox committed:
Feat: Updatd Auto Balance to not overshoot

GearBoxFox committed:
Feat: Updatd AutoFactory to have balance options

GearBoxFox committed:
Feat: new auto balance control style

GearBoxFox committed:
Merge branch 'GrandCircuit-PathplanningTest' of https://github.com/TitaniumTitans/2023ChargedUp into GrandCircuit-PathplanningTest

GearBoxFox committed:
Feat: Single cube auto

GearBoxFox committed:
Merge remote-tracking branch 'origin/Gearfox-DriveBindings' into GrandCircuit-PathplanningTest

# Conflicts:
#	src/main/java/frc/robot/subsystems/swerve/SwerveDrivetrain.java

Zoarial94 commented:
@GearBoxFox instead of having manual position control while balancing, can you have the path planner keep the current position?

Zoarial94 commented on src/main/java/frc/robot/commands/AutoBalance.java:
```diff
@@ -52,35 +61,53 @@ public void execute() {
     m_drivePower = Math.min(m_balanceController.calculate(m_currentAngle), 1);
     // Limit max power
-    m_drivePower = MathUtil.clamp(m_drivePower, -0.6, 0.6);
+    m_drivePower = MathUtil.clamp(m_drivePower, -0.5, 0.5);
```
Math.min isn't doing anything here.

GearBoxFox replied:
This is the older version, this isn't being used anymore. The newer balance command is AutoBalanceTransCommand.

Zoarial94 commented on src/main/java/frc/robot/commands/AutoBalance.java:
```diff
@@ -52,35 +61,53 @@ public void execute() {
+      m_balanceTimer.start();
+      m_drivePower = Math.copySign(0.15, -m_drive.getGyroPitch().getDegrees());
+    }
+
+    if (m_balanceTimer.hasElapsed(0.01)) {
+//      m_drivePower = Math.copySign(0.15, -m_drive.getGyroPitch().getDegrees());
       m_drivePower = 0.0;
-//      SmartDashboard.putNumber("Stop Balancing", 1);
+    }
+    if (m_balanceTimer.hasElapsed(0.75)) {
+      m_balanceTimer.stop();
+      m_balanceTimer.reset();
     }
 
     // Counter for checking if robot is truly balanced
-    if (Math.abs(m_currentAngle) < AutoConstants.DESIRED_BALANCE_ANGLE) {
+    if (Math.abs(m_error) < AutoConstants.ACCEPTABLE_BALANCE_ANGLE && !m_timer.hasElapsed(0.1)) {
       // if robot is balanced, stop moving and start a timer
       m_timer.start();
       m_drivePower = 0.0;
-    } else if (Math.abs(m_currentAngle) < AutoConstants.DESIRED_BALANCE_ANGLE && m_timer.get() >= 10) {
+      SmartDashboard.putBoolean("Timer Started", true);
+    }
+    if (Math.abs(m_error) < AutoConstants.ACCEPTABLE_BALANCE_ANGLE && m_timer.hasElapsed(2)) {
       // If robot has been balanced for 10 seconds, stop the command
       m_timer.stop();
       m_isLevel = true;
-    } else if (m_timer.get() > 0 && Math.abs(m_currentAngle) > AutoConstants.DESIRED_BALANCE_ANGLE) {
+      SmartDashboard.putBoolean("Timer Started", false);
+    }
+    if (Math.abs(m_error) < AutoConstants.ACCEPTABLE_BALANCE_ANGLE && m_timer.get() < 2) {
+      m_drivePower = 0.0;
+    }
+    if (m_timer.hasElapsed(0) && Math.abs(m_error) > AutoConstants.ACCEPTABLE_BALANCE_ANGLE) {
```
These if statements should be mutually exclusive. You don't want unknowingly take multiple paths. You may need to reorder the statements as well to get the correct behavior.

GearBoxFox replied:
Ok, I think we should merge the PR before we resolve this, as like i said earlier this version of the command isn't being used anymore.

Zoarial94 commented on src/main/java/frc/robot/subsystems/swerve/SwerveDrivetrain.java:
```diff
@@ -204,8 +204,12 @@ public Rotation2d getGyroPitch() {
+    }
+
+    public double getGyroPitchRate() {
+        return m_currentPitch - m_previousPitch;
```
This isn't really the pitch rate unless it accounts for time.

GearBoxFox replied:
I can update the method to take time into account

Zoarial94 committed:
Constants.java cleanup

Zoarial94 committed:
SwerveDrivetrain.java cleanup

Zoarial94 committed:
AutoBalance.java cleanup

Zoarial94 commented on src/main/java/frc/robot/commands/AutoBalance.java:
```diff
@@ -52,35 +61,53 @@ public void execute() {
     // if robot is tilting forward, stop moving so we don't overshoot the balance
     if (Math.abs(rollRateBalance) >= 0.3) {
-//      m_drivePower = Math.copySign(0.0, -m_drive.getGyroPitch().getDegrees());
+      m_balanceTimer.start();
+      m_drivePower = Math.copySign(0.15, -m_drive.getGyroPitch().getDegrees());
+    }
+
+    if (m_balanceTimer.hasElapsed(0.01)) {
+//      m_drivePower = Math.copySign(0.15, -m_drive.getGyroPitch().getDegrees());
       m_drivePower = 0.0;
-//      SmartDashboard.putNumber("Stop Balancing", 1);
+    }
+    if (m_balanceTimer.hasElapsed(0.75)) {
+      m_balanceTimer.stop();
+      m_balanceTimer.reset();
```
This shouldn't be timer based, it should be based off the robot pitch movement.

GearBoxFox replied:
That can be done.

Zoarial94 committed:
Merged some cleaned up code into GrandCircuit-PathplanningTest

GearBoxFox committed:
Feat: one cone-buce auto and new intake motor

GearBoxFox committed:
Feat: tuning of auto align and arm setpoints plus maintnance mode

GearBoxFox committed:
Merge remote-tracking branch 'origin/GrandCircuit-PathplanningTest' into GrandCircuit-PathplanningTest

# Conflicts:
#	src/main/java/frc/robot/Constants.java
#	src/main/java/frc/robot/subsystems/swerve/SwerveDrivetrain.java

GearBoxFox committed:
Fix: merged origin and resolved conflicts

GearBoxFox committed:
Feat: One piece and engage auto

GearBoxFox committed:
Feat: tuned one piece auto and fixed arm ext at setpoint error

Zoarial94 committed:
Added Pickup Middle auto. Minor cleanup.

For arm subsystems, renamed the previously verbose at-setpoint function to just atSetpoint()

Zoarial94 committed:
Pickup Middle auto can now score middle or high. Adjusted path.

Zoarial94 committed:
FEAT: Controller will rumble when intaking and the motor stalls

Zoarial94 committed:
Formatting and cleanup

GearBoxFox committed:
Feat: two piece auto?

GearBoxFox committed:
Merge branch 'GrandCircuit-PathplanningTest' of https://github.com/TitaniumTitans/2023ChargedUp into GrandCircuit-PathplanningTest

Zoarial94 committed:
Updated paths

GearBoxFox committed:
Fix: tuned path

GearBoxFox committed:
Merge branch 'GrandCircuit-PathplanningTest' of https://github.com/TitaniumTitans/2023ChargedUp into GrandCircuit-PathplanningTest

Zoarial94 committed:
Fixed rumble on intake.

GearBoxFox committed:
Feat: pitch rate take in account of time

GearBoxFox committed:
Feat: linter fix of SparkMaxFactory.kt

GearBoxFox committed:
Fix: More linting changes

GearBoxFox committed:
Final github linter fixes

Zoarial94 commented on src/main/java/frc/robot/subsystems/swerve/SwerveDrivetrain.java:
```diff
@@ -196,16 +210,25 @@ public void setModuleStates(SwerveModuleState[] states) {
+    }
+
+    public double getGyroPitchRate() {
+        double fpgaElapsedTime = RobotController.getFPGATime() - m_prevTime;
```
We need to handle the edge case where m_prevTime == 0. If it is zero, set m_prevTime to the current time.

Zoarial94 commented on src/main/java/frc/robot/subsystems/swerve/SwerveDrivetrain.java:
```diff
@@ -196,16 +210,25 @@ public void setModuleStates(SwerveModuleState[] states) {
+
+    public double getGyroPitchRate() {
+        double fpgaElapsedTime = RobotController.getFPGATime() - m_prevTime;
+        return (m_currentPitch - m_previousPitch) / RobotController.getFPGATime();
```
Divide by the difference.

Zoarial94 commented on src/main/java/frc/robot/commands/AutoBalance.java:
```diff
@@ -41,46 +38,71 @@ public AutoBalance(SwerveDrivetrain drive) {
+    m_drivePower = MathUtil.clamp(m_drivePower, -0.5, 0.5);
 
-    double rollRateBalance = m_drive.getGyroRollPitch();
+    double rollRateBalance = m_drive.getGyroPitchRate();
```
The number returned here is different now that it is adjusted for time elapsed.

GearBoxFox replied:
wouldn't it still be the rate of the roll, just properly calculated now as we're accouting for time elapsed?

Zoarial94 replied:
Correct. Just making a note for later.

GearBoxFox committed:
Feat: Time enabled counter

GearBoxFox committed:
Fix: gyro roll rate uses difference in time and accounted for bootup timer

GearBoxFox committed:
Feat: no more janky command logic for intake control

GearBoxFox commented:
@Zoarial94 I finished those last two things you requested, am I good to merge into main?

Zoarial94 reviewed (approved):
Looks good. We need to test ASAP though.

GearBoxFox commented:
Yes, we can run another system check when we get to the pit.
